dns_cache/manager.go

```diff
@@ -0,0 +1,62 @@
+package dns_cache
+
+import (
+	"context"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/TykTechnologies/tyk/log"
+)
+
+var (
+	logger = log.Get().WithField("prefix", "dns-cache")
+)
+
+type dialContextFunc func(ctx context.Context, network, address string) (net.Conn, error)
+
+type IDnsCacheManager interface {
+	InitDNSCaching(ttl, checkInterval time.Duration)
+	WrapDialer(dialer *net.Dialer) dialContextFunc
+}
+
+type DnsCacheManager struct {
+	DnsCache *DnsCacheStorage
+}
+
+func NewDnsCacheManager() *DnsCacheManager {
+	return &DnsCacheManager{nil}
+}
+
+func (m *DnsCacheManager) WrapDialer(dialer *net.Dialer) dialContextFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
+		return m.doCachedDial(dialer, ctx, network, address)
+	}
+}
+
+func (m *DnsCacheManager) doCachedDial(d *net.Dialer, ctx context.Context, network, address string) (net.Conn, error) {
+	if m.DnsCache == nil {
```

Comment:
might worth to add one more short path to check if address already contains IP address (in some cases upstream might be specified by IP for host name) and return right away `return d.DialContext(ctx, network, address)`

This would make it slightly faster and we won't cache items like key,value->IP,IP